File: 백준/Platinum/17071. 숨바꼭질 5/숨바꼭질 5.cc
#include<bits/stdc++.h>
using namespace std;
const int mx = 500000;
int visited[2][mx + 4], n, k, ok, turn = 1;
int main(){
    cin >> n >> k;
    if(n == k){cout << 0 << '\n'; return 0;}
    queue<int> q;
    visited[0][n] = 1; // 왜2차원 배열을 선언했을까?
    q.push(n);
    while(q.size()){
        k += turn;
        if(k > mx) break;
        if(visited[turn % 2][k]){ // 시간 종료조건
            ok = true;
            break;
        }
        int qSize = q.size(); // 시간과 탐색의 횟수를 맞춰줄려고
        for(int i = 0; i < qSize; i++){
            int now = q.front(); q.pop();
            for(int next : {now + 1, now - 1, now * 2}){
                if(next < 0 || next > mx) continue;
                if(visited[turn % 2][next]) continue;
                visited[turn % 2][next] = visited[(turn + 1) % 2][next] + 1;
                if(next == k){ // bfs탐색 종료조건
                    ok = 1;
                    break;
                }
                q.push(next);
            }
            if(ok) break; // 탐색종료
        }
        if(ok) break; // 탐색종료
        turn++;
    }
    if(ok) cout << turn << '\n'; // ok
    else cout << -1 << '\n';
    return 0;
}

File: 백준/Platinum/17071. 숨바꼭질 5/숨바꼭질 5.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    static final int mx = 500000;
    static int[][] visited = new int[2][mx + 4];
    static int n, k, turn = 1;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        k = sc.nextInt();
        boolean ok = false;

        if (n == k) {
            System.out.println(0);
            return;
        }

        Queue<Integer> q = new LinkedList<>();
        visited[0][n] = 1;
        q.add(n);

        while (!q.isEmpty()) {
            k += turn;
            if (k > mx) break;
            if (visited[turn % 2][k] != 0) {
                ok = true;
                break;
            }

            int qSize = q.size();
            for (int i = 0; i < qSize; i++) {
                int now = q.poll();
                for (int next : new int[] { now + 1, now - 1, now * 2 }) {
                    if (next < 0 || next > mx) continue;
                    if (visited[turn % 2][next] != 0) continue;
                    visited[turn % 2][next] = visited[(turn + 1) % 2][next] + 1;
                    if (next == k) {
                        ok = true;
                        break;
                    }
                    q.add(next);
                }
                if (ok) break;
            }
            if (ok) break;
            turn++;
        }

        if (ok) System.out.println(turn);
        else  System.out.println(-1);
    }
}
